### PR TITLE
popup: closes when user clicks button or some inputs via spacebar (close #1134)

### LINKED ISSUE #1134: popup: closes when user clicks button or some inputs via spacebar
The native `<input />` with some types (`button`, `reset`, `submit`, `image`, `checkbox`, `radio`, `file`) and  the `<button></button>` generate `click` event when user press spacebar on it. This click bubbles  to document and closes all popups

### PR DESCRIPTION
**Pull request by persidskiy**
close #1134
// cc @narqo @dfilatov

### CODE REVIEW

File: common.blocks/popup/popup.js [hunk 1]
```diff
@@ -45,6 +46,7 @@ provide(BEMDOM.decl(this.name, /** @lends popup.prototype */{
                 this
                     ._captureZIndex()
                     ._bindToParentPopup()
+                    .bindTo('keydown', this._onKeyDown)
                     .bindTo('pointerpress', this._onPointerPress);
```
Comment (narqo) [inline common.blocks/popup/popup.js]: Couldn't we add `pointerclick` listener here as well?

File: common.blocks/popup/popup.js [hunk 2]
```diff
@@ -76,10 +79,23 @@ provide(BEMDOM.decl(this.name, /** @lends popup.prototype */{
         return res;
     },
 
+    _onKeyDown : function(e) {
+        var nodeName = e.target.nodeName.toLowerCase();
+
+        e.keyCode === keyCodes.SPACE && (
+            nodeName === 'button' ||
+            (nodeName === 'input' && ~INPUT_TYPES_TRIGGER_CLICK.indexOf($(e.target).attr('type')))
```
Comment (narqo) [inline common.blocks/popup/popup.js]: May be we should add this to `bem-core/dom`?

File: common.blocks/popup/popup.js (hunk 1 above)
Comment (persidskiy) [inline common.blocks/popup/popup.js]: No, https://github.com/bem/bem-components/pull/900/files

File: common.blocks/popup/popup.js (hunk 2 above)
Comment (persidskiy) [inline common.blocks/popup/popup.js]: I think it's very specific. At least, I was faced with it only here.

File: common.blocks/popup/popup.js (hunk 1 above)
Comment (narqo) [inline common.blocks/popup/popup.js]: I'm not talking about removing `pointerpress`, but about adding `pointerclick` listener for cases where `click` event generated programmatically. It seems it should solve our problem in more general way. 

Of course current `_onPointerPress` should have some small changes too (and not to run two `while` loops for ppress and click).

Comment (persidskiy) [inline common.blocks/popup/popup.js]: Hmm, it's good idea! :)

**Comment by persidskiy** [comment]
:arrow_up:

File: common.blocks/popup/popup.js [hunk 3]
```diff
@@ -76,10 +76,10 @@ provide(BEMDOM.decl(this.name, /** @lends popup.prototype */{
         return res;
     },
 
-    _onPointerPress : function() {
+    _preventHideByClick : function() {
         var curPopup = this;
         do {
-            curPopup._inPopupPointerPress = true;
+            curPopup._needPreventHideByClick = true;
         } while(curPopup = curPopup._getParentPopup());
```
Comment (narqo) [inline common.blocks/popup/popup.js:83]: Maybe we could optimise `_preventHideByClick` method in some way? Now this loop runs twice on each click: first while handling `pointerpress`, second on `pointerclick`.

Comment (persidskiy) [inline common.blocks/popup/popup.js:83]: Yep, but it shouldn't be expensive. And it more robust.

File: common.blocks/popup/_autoclosable/popup_autoclosable.js [hunk 4]
```diff
@@ -41,8 +41,8 @@ provide(Popup.decl({ modName : 'autoclosable', modVal : true }, /** @lends popup
         if(this.hasMod('target', 'anchor') && dom.contains(this._anchor, $(e.target)))
             return;
 
-        this._inPopupPointerPress?
-           this._inPopupPointerPress = null :
+        this._needPreventHideByClick?
```
Comment (dfilatov) [inline common.blocks/popup/_autoclosable/popup_autoclosable.js]: Use either `_preventHideByClick` or `_needToPreventHideByClick`. I like first one.

**Comment by persidskiy** [comment]
@dfilatov 

```
_preventHideByClick() -> _setPreventHideByClick()
_needPreventHideByClick -> _preventHideByClick
```

**Comment by dfilatov** [comment]
@Coltspb ok

**Comment by persidskiy** [comment]
Then, what about merge?

**Comment by dfilatov** [comment]
lgtm

**Comment by persidskiy** [comment]
@narqo ?

**Comment by narqo** [comment]
:+1:

**Pull request merged**